spec/declaration/tuple_spec.lua

```diff
@@ -48,6 +52,7 @@ describe("tuple declarations", function()
    pending("should error with explicit integer indices that are out of range", util.check_type_error([[
       local c: {number, string} = { [-1] = 10 }
    ]], {
-      { msg = "in local declaration: c: got {number : number}, expected {number, string}" },
+      -- Something about the table being a map or about -1 being out of range
+      -- { msg = "in local declaration: c: got {number : number}, expected {number, string}" },
```

Comment:
What does the code above produce, with this PR?

Reply:
It currently produces the error
`'in local declaration: c: tuple entry 2 of type string does not match type of array elements, which is number'` which seems fine given that that's inferred as `{number}`. I think I marked that as pending because of how i wanted to change tables of that form to maps rather than arrays in the original pr.

Reply:
I definitely would prefer something saying something like `-1 is out of range for the tuple {number, string}`, but I think that would require types inferred from literals to carry more information than they currently do iirc

Reply:
> i wanted to change tables of that form to maps rather than arrays in the original pr.

Ah yes, I changed that bit of "infer as map" but ended up not discussing it much. I'm generally trying to avoid surprises to Lua programmers by not assigning two different semantic meanings to two syntactic constructs that happen to be syntax-sugar to each other in Lua. (But I did break that guideline on occasion: function declarations with `:` keep the "I'm a method" metadata and are treated differently).

Reply:
I'll remove this test case when merging the PR, because it's equivalent to the `c` case in line 26 :+1: